src/relier/OAuthUtils.js

```diff
@@ -0,0 +1,159 @@
+const fxaKeyUtils = new KeyUtils();
+
+const OAUTH_SERVER_URL = 'https://oauth.accounts.firefox.com/v1';
+
+/**
+ * OAuth methods to authenticate with Firefox Accounts and get scoped keys
+ * @module relier-OAuthUtils
+ * @example
+ * ```js
+ * const fxaKeysUtil = new fxaCryptoRelier.OAuthUtils();
+ *
+ * fxaKeysUtil.launchWebExtensionKeyFlow('YOUR_CLIENT_ID', {
+ *   redirectUri: browser.identity.getRedirectURL(),
+ *   scopes: ['profile', 'https://identity.mozilla.org/apps/lockbox'],
+ * }).then((loginDetails) => {
+ *   const key = loginDetails.keys['https://identity.mozilla.org/apps/lockbox'];
+ *   const credentials = {
+ *     access_token: loginDetails.access_token,
+ *     refresh_token: loginDetails.refresh_token,
+ *     key
+ * };
+ * ```
+ */
+class OAuthUtils {
+  /**
+   * @constructor
+   * @param {object} [options]
+   * @param {string} [options.oauthServer] - Custom Firefox Accounts OAuth server
+   */
+  constructor(options = {}) {
+    this.oauthServer = options.oauthServer || OAUTH_SERVER_URL;
+  }
+  /**
+   * @method launchWebExtensionKeyFlow
+   * @desc Used to launch the Firefox Accounts scope key login flow in WebExtensions
+   * @param {string} clientId - FxA relier client id
+   * @param {object} [options={}]
+   * @param {array} [options.scopes=[]] - Requested OAuth scopes
+   * @param {object} [options.browserApi=browser] - Custom browser API override
+   * @param {function} [options.getBearerTokenRequest=getBearerTokenRequest] -
+   *   Custom getBearerTokenRequest function override
+   * @returns {Promise}
+   */
+  launchWebExtensionKeyFlow(clientId, options = {}) {
+    if (! clientId) {
+      throw new Error('clientId required');
+    }
+
+    const browserApi = options.browserApi || browser;
+    const getBearerTokenRequest = options.getBearerTokenRequest || this._getBearerTokenRequest;
+    const SCOPES = options.scopes || [];
+
+    const state = util.createRandomString(16);
+    const codeVerifier = util.createRandomString(43);
+    const queryParams = {
+      access_type: 'offline', // eslint-disable-line camelcase
+      client_id: clientId, // eslint-disable-line camelcase
+      redirect_uri: options.redirectUri, // eslint-disable-line camelcase
+      scope: SCOPES.join(' '),
+      state: state
+    };
+
+    return util.sha256base64url(codeVerifier).then(codeChallenge => {
+      queryParams.response_type = 'code'; // eslint-disable-line camelcase
+      queryParams.code_challenge_method = 'S256'; // eslint-disable-line camelcase
+      queryParams.code_challenge = codeChallenge; // eslint-disable-line camelcase
+
+      return fxaKeyUtils.createApplicationKeyPair();
+    }).then((keyTypes) => {
+      const base64JwkPublicKey = jose.util.base64url.encode(JSON.stringify(keyTypes.jwkPublicKey), 'utf8');
+
+      queryParams.keys_jwk = base64JwkPublicKey; // eslint-disable-line camelcase
+
+      const authUrl = `${this.oauthServer}/authorization` + util.objectToQueryString(queryParams);
```

Comment:
Let's add a follow-up bug about discovering this URI by looking up https://accounts.firefox.com/.well-known/openid-configuration; I know we can't do that right now because of CORS, but I feel pretty strongly about that being the Right Thing in the longer term.

Reply:
https://github.com/mozilla/fxa-content-server/issues/5453